// search-client/es/src/Common/QueryChangeSpecifications.js
/**
 * Defines the enum bit-field flags that signifies which query-fields that may resolve in changed results (both Categorize and Find).
 */
export var QueryChangeSpecifications;
(function (QueryChangeSpecifications) {
    QueryChangeSpecifications[QueryChangeSpecifications["none"] = 0] = "none";
    QueryChangeSpecifications[QueryChangeSpecifications["clientId"] = 1] = "clientId";
    QueryChangeSpecifications[QueryChangeSpecifications["categorizationType"] = 2] = "categorizationType";
    QueryChangeSpecifications[QueryChangeSpecifications["dateFrom"] = 4] = "dateFrom";
    QueryChangeSpecifications[QueryChangeSpecifications["dateTo"] = 8] = "dateTo";
    QueryChangeSpecifications[QueryChangeSpecifications["filters"] = 16] = "filters";
    QueryChangeSpecifications[QueryChangeSpecifications["matchGenerateContent"] = 32] = "matchGenerateContent";
    QueryChangeSpecifications[QueryChangeSpecifications["matchGenerateContentHighlights"] = 64] = "matchGenerateContentHighlights";
    QueryChangeSpecifications[QueryChangeSpecifications["matchGrouping"] = 128] = "matchGrouping";
    QueryChangeSpecifications[QueryChangeSpecifications["matchOrderBy"] = 256] = "matchOrderBy";
    QueryChangeSpecifications[QueryChangeSpecifications["matchPage"] = 512] = "matchPage";
    QueryChangeSpecifications[QueryChangeSpecifications["matchPageSize"] = 1024] = "matchPageSize";
    QueryChangeSpecifications[QueryChangeSpecifications["maxSuggestions"] = 2048] = "maxSuggestions";
    QueryChangeSpecifications[QueryChangeSpecifications["queryText"] = 4096] = "queryText";
    QueryChangeSpecifications[QueryChangeSpecifications["searchType"] = 8192] = "searchType";
    QueryChangeSpecifications[QueryChangeSpecifications["uiLanguageCode"] = 16384] = "uiLanguageCode";
})(QueryChangeSpecifications || (QueryChangeSpecifications = {}));
